# STAT&BASIC-4. Тестирование гипотез 
## Юнит 2. Разведывательный анализ данных
### Skillfactory: DST-10

In [24]:
import pandas as pd
import numpy as np
from scipy.stats import norm, t
import matplotlib.pyplot as plt

## Задание 4.7.1  
Условие:  
Производитель аккумуляторных батарей заявляет, что средний срок службы батарейки АА составляет 14 месяцев. Вы подозреваете, что на самом деле это число преувеличено.  
Чтобы проверить утверждение производителя, вы взяли случайную выборку из 64 батареек и посчитали выборочное среднее — 13.5 месяцев. Известно, что срок службы батарейки имеет стандартное отклонение 2 месяца. Для тестирования гипотезы используйте уровень значимости (alpha =0.05).  
Обратите внимание, что как постановка альтернативной гипотезы, так и выбор уровня значимости alpha может повлиять на вывод, который вы делаете в результате тестирования гипотезы.

1. Первый шаг при тестировании – постановка нулевой и альтернативной гипотез. Нулевая гипотеза: μ=14. Какую альтернативную гипотезу нужно использовать в данном случае?  

Ответ: μ<14

In [25]:
# создадим функцию для z z-статистики
def z_statistic(E, mu_0, sigma, n):
    return (E-mu_0)/(sigma/(n**0.5))

## Задание 4.7.2
Чему равна z-статистика?

In [26]:
n = 64  # элементов в выборке
E_sampling_64 = 13.5  # матожидание выборки

sigma = 2  # стандартное отклонение 
mu_0 = 14

z_stat = z_statistic(E_sampling_64, mu_0, sigma, n)

# проверяем допущения центральной предельной теоремы:
# Случайна ли выборка? Да.
# Достаточно ли велика выборка? да n>30  

task = "4.7.2"
print(f"Ответ к заданию {task}")
print(f"z-статистика равна:= {round(z_stat,0)}")

Ответ к заданию 4.7.2
z-статистика равна:= -2.0


## Задание 4.7.3
Чему равно значение p-value (ответ округлите до 3 знаков после разделительной точки)?

In [27]:
p_value = norm.cdf(z_stat)

task = "4.7.3"
print(f"Ответ к заданию {task}")
print(f"значение p-value равно:= {round(p_value,3)}")

Ответ к заданию 4.7.3
значение p-value равно:= 0.023


## Задание 4.7.4

P-value (0.023) оказалась меньше alpha (0.05), что дает нам право отвергнуть нулевую гипотезу в пользу альтернативной:  

Ответ: Таким образом, мы ***отвергаем*** нулевую гипотезу и приходим к выводу, что средний срок службы батарейки ***меньше*** 14 мес.

## Задание 4.7.5  
Условие:
На упаковке муки указана ее масса — 455 г. Известно, что масса муки в упаковке распределена приблизительно нормально, и что ее стандартное отклонение составляет 12.5 граммов. В случайной выборке из 16 упаковок средняя масса составила 449 граммов.

Протестируйте на уровне значимости 5% нулевую гипотезу о том, что средняя масса упаковки с мукой составляет 455 граммов.  

Допустим, у вас нет никаких догадок о том, является ли истинная средняя масса упаковки  выше или ниже указанного значения в  граммов. Сформулируйте соответствующую альтернативную гипотезу и протестируйте ее.  
Чему равно соответствующее значение p-value? Ответ округлите до 3 знаков после разделительной точки.

In [28]:
n = 16  # элементов в выборке
E_sampling_16 = 449  # матожидание выборки

sigma = 12.5  # стандартное отклонение 
mu_0 = 455

z_stat = z_statistic(E_sampling_16, mu_0, sigma, n)
p_value = 2 * norm.cdf(z_stat)  # умножаем на 2 потому что в алтернативной гипотезе 
                                # условие !=, это означает что нам надо найти P(|Z|>z_stat)

# проверяем допущения центральной предельной теоремы:
# Случайна ли выборка? Да.
# Достаточно ли велика выборка? НЕТ n<30 Но в условии - Известно, что масса муки в упаковке
# распределена приблизительно нормально. Поэтому пренебрегаем проверкой на величину выборки

task = "4.7.5"
print(f"Ответ к заданию {task}")
print(f"p_value равно:= {round(p_value,3)}")

Ответ к заданию 4.7.5
p_value равно:= 0.055


## Задание 4.7.7  
Вы заподозрили, что производитель сознательно завышает количество граммов, указанное на упаковке, в то время как на самом деле средняя масса ниже. Сформулируйте соответствующую альтернативную гипотезу и протестируйте ее.  
Чему в таком случае равно значение p-value? Ответ округлите до 3 знаков после разделительной точки.

In [29]:

p_value = norm.cdf(z_stat)  

# проверяем допущения центральной предельной теоремы:
# Случайна ли выборка? Да.
# Достаточно ли велика выборка? НЕТ n<30 Но в условии - Известно, что масса муки в упаковке
# распределена приблизительно нормально. Поэтому пренебрегаем проверкой на величину выборки

task = "4.7.7"
print(f"Ответ к заданию {task}")
print(f"p_value равно:= {round(p_value,3)}")

Ответ к заданию 4.7.7
p_value равно:= 0.027


## Задание 4.8.1
Условие:
Производитель утверждает, что новая модель кондиционера для воздуха потребляет только 6.5 кВт⋅ч электричества в день. Однако в союзе потребителей убеждены, что истинное среднее количество потребляемой энергии выше этого значения.  
Союз проводит собственное исследование на основании случайной выборки в 25 кондиционеров и находит, что выборочное среднее значение равно 7 кВт⋅ч. Известно, что количество потребляемой кондиционером энергии распределено нормально и его истинное стандартное отклонение равно 1.4 кВт⋅ч. 

Какую тестовую статистику следует использовать в данном случае?  
Ответ: z-статистику

## Задание 4.8.2  
Введите значение p-value, округлив до 3 знаков после разделительной точки.

In [30]:
n = 25  # элементов в выборке
E_sampling_25 = 7  # матожидание выборки

sigma = 1.4  # стандартное отклонение 
mu_0 = 6.5

z_stat = z_statistic(E_sampling_25, mu_0, sigma, n)
p_value = 1 - norm.cdf(z_stat)  # если мы ищем P(X<z_stat|H_zero) = P(Z<z_stat)
                                # P(Z>z_stat) = norm.cdf(z_stat)
                                # если мы ищем P(X>z_stat|H_zero) = P(Z>z_stat)
                                # P(Z>z_stat) = 1 - norm.cdf(z_stat)

# проверяем допущения центральной предельной теоремы:
# Случайна ли выборка? Да.
# Достаточно ли велика выборка? НЕТ n<30 Но в условии - Известно, что количество 
# потребляемой кондиционером энергии распределено нормально. 
# Поэтому пренебрегаем проверкой на величину выборки

task = "4.8.2"
print(f"Ответ к заданию {task}")
print(f"p_value равно:= {round(p_value,3)}")

Ответ к заданию 4.8.2
p_value равно:= 0.037


## Задание 4.8.4  
Региональные власти утверждают, что средний годовой доход домохозяйства составляет **$12000**.  
Экономист, подозревая, что истинный средний доход может быть меньше, находит на случайной выборке из 135 домохозяйств этого региона, что средний доход равен **$11500**,  
а стандартное отклонение равно **$5110**. 
 
Укажите значение тестовой статистики, которую следует использовать в данном случае, округляя до 3 знаков после разделительной точки.

In [31]:
n = 135  # элементов в выборке
E_sampling_135 = 11500  # матожидание выборки

s = 5110  # стандартное отклонение выборки
mu_0 = 12000

t_stat = z_statistic(E_sampling_135, mu_0, s, n)

task = "4.8.4"
print(f"Ответ к заданию {task}")
print(f"тестовая статистика равна:= {round(t_stat,3)}")

Ответ к заданию 4.8.4
тестовая статистика равна:= -1.137


## Задание 4.8.5
Введите значение p-value, округлив до 3 знаков после разделительной точки.

In [32]:
p_value = t.cdf(t_stat, n-1)  
# Чтобы найти p-value для t-статистики
# t.cdf(), если в альтернативной гипотезе стоит знак <; эта функция находит площадь 
# под «левым хвостом» на графике плотности распределения Стьюдента — площадь под кривой
# слева от заданной точки;
# 1 - t.cdf(), если в альтернативной гипотезе стоит знак >; эта функция находит площадь 
# под «правым хвостом» на графике плотности распределения Стьюдента — площадь под
# кривой справа от заданной точки. 
# Вместо этой функции можно использовать метод t.sf() = 1 - t.cdf().
# (1 - t.cdf()) * 2, если в альтернативной гипотезе стоит знак ≠; эта функция считает 
# площадь под правым хвостом. Первый аргумент берем по модулю 2*(1 -t.cdf(abs(x), df).

# проверяем допущения центральной предельной теоремы:
# Случайна ли выборка? Да.
# Достаточно ли велика выборка? Да n>30 

task = "4.8.5"
print(f"Ответ к заданию {task}")
print(f"p_value равно:= {round(p_value,3)}")

Ответ к заданию 4.8.5
p_value равно:= 0.129


!!! внимание на платформе ошибка. Правильный ответ в системе 0.128. На самом деле - Расчетное значение 0.12881, которое при окреглении дает 0.129 

## Задание 4.8.7
Условие:  
Сотрудники компании, специализирующейся на продаже франшиз, заявляют, что процент возврата на инвестиции в проданные ими франшизы уже в первый год составляет 10 (все дальнейшие значения приведены в процентах, их можно игнорировать как единицы измерения). Была взята выборка из 10 различных сделок данной компании и рассчитан процент возврата на инвестиции в первый год в каждой из них. Значения приведены в коде ниже.  
Вы хотите проверить достоверность заявления, сделанного компанией, подозревая, что на самом деле средний процент возврата на инвестиции в первый год ниже 10.  
Чтобы протестировать соответствующую гипотезу, сначала рассчитайте в Python значение среднего и стандартного отклонения.  
Какую тестовую статистику следует использовать в данном случае?  

**Ответ:**  t-статистику

In [33]:
X = np.array([6.1, 9.2, 11.5, 8.6, 12.1, 3.9, 8.4, 10.1, 9.4, 8.9])
X_mean = X.mean()
X_std = X.std(ddof=1)  # различия в расчетах std в numpy и в pandas. 
# Дело в том, что Pandas использует несмещенную оценку (N-1 в знаменателе), 
# а Numpy по умолчанию нет. Чтобы получить у Numpy тот же метод расчета (и результат в итоге)
# передайте аргумент  ddof=1 в numpy.std().
print(f"среднее := {X_mean}, среднее :={X_std}")

n = 10  # элементов в выборке
E_sampling_10 = X_mean  # матожидание выборки

s = X_std  # стандартное отклонение выборки
mu_0 = 10

t_stat = z_statistic(E_sampling_10, mu_0, s, n)
p_value = t.cdf(t_stat, n-1)  
# Чтобы найти p-value для t-статистики
# t.cdf(), если в альтернативной гипотезе стоит знак <; эта функция находит площадь 
# под «левым хвостом» на графике плотности распределения Стьюдента — площадь под кривой
# слева от заданной точки;
# 1 - t.cdf(), если в альтернативной гипотезе стоит знак >; эта функция находит площадь 
# под «правым хвостом» на графике плотности распределения Стьюдента — площадь под
# кривой справа от заданной точки. 
# Вместо этой функции можно использовать метод t.sf() = 1 - t.cdf().
# (1 - t.cdf()) * 2, если в альтернативной гипотезе стоит знак ≠; эта функция считает 
# площадь под правым хвостом. Первый аргумент берем по модулю 2*(1 -t.cdf(abs(x), df).

# проверяем допущения центральной предельной теоремы:
# Случайна ли выборка? Да.
# Достаточно ли велика выборка? Да n>30 
task = "4.8.7"
print(f"Ответ к заданию {task}")
print(f"p_value равно:= {round(p_value,3)}")

среднее := 8.820000000000002, среднее :=2.401295946405237
Ответ к заданию 4.8.7
p_value равно:= 0.077


## Задание 4.8.8  
Найдите p-value и сделайте вывод о заявлении компании, используя уровень значимости 5 %.  

**Ответ:** p-value>0.05, поэтому мы не можем опровергнуть заявление компании верно

## Задание 4.8.9  
Какие предположения должны быть выполнены для использования данного статистического теста? Выберите все варианты, которые подходят:  

**Ответ:**
- Случайная выборка
- Распределение возврата на инвестиции приблизительно нормальное

## Задание 4.10.1  
Условие:  
Представитель профсоюза утверждает, что 75% сотрудников готовы присоединиться к забастовке, если их требования не будут удовлетворены. Сотрудник компании, ведущий переговоры с профсоюзом, считает, что эта пропорция сильно преувеличена. Он решает проверить утверждение представителя профсоюза, используя уровень значимости 5%.  
Таким образом, ему надо протестировать следующий набор гипотез: 

H~0~ : p = 0.75, H~A~ : p < 0.75

В выборке из 125 сотрудников 87 ответили, что будут готовы поддержать забастовку в случае неудовлетворения требования профсоюза.  
Чему равна z-статистика для данного теста? Ответ округлите до 3 знаков после разделительной точки.

In [34]:
p_0 = 0.75  # Мы хотим проверить, не снизилась ли уже наша пропорция ниже 0.75

# cобираем наблюдения
p_ = 87/125
n = 125

# cтроим наши ожидания. 
# выборка случайная
# достаточно большая: n*p_==87 > 5
n*(1-p_)

38.00000000000001

In [38]:
# n*(1-p_) > 5

# Предполагаем, что нулевая гипотеза верна и заменяем p на p_0:
z_stat_prop = (p_ - p_0)/(((p_0*(1-p_0))/n)**0.5)

task = "4.10.1"
print(f"Ответ к заданию {task}")
print(f"z-статистика для данного теста равна:= {round(z_stat_prop,3)}")

Ответ к заданию 4.10.1
z-статистика для данного теста равна:= -1.394


## Задание 4.10.2  
Чему равно значение p-value? Ответ округлите до 3 знаков после разделительной точки.

In [39]:
p_value = norm.cdf(z_stat_prop)

task = "4.10.2"
print(f"Ответ к заданию {task}")
print(f"p_value равно:= {round(p_value,3)}")

Ответ к заданию 4.10.2
p_value равно:= 0.082


## Задание 4.10.3  
К какому выводу придет сотрудник компании?

**Ответ:** Возможно, представитель профсоюза действительно прав и пропорция сотрудников, готовых присоединиться к забастовке равна 75% верно

конец модуля